runtime/sema/check_member_expression.go

```diff
@@ -376,7 +376,9 @@ func (checker *Checker) visitMember(expression *ast.MemberExpression, isAssignme
 // in the current location of the checker, along with the authorzation with which the result can be used
 func (checker *Checker) isReadableMember(accessedType Type, member *Member, resultingType Type, accessRange func() ast.Range) (bool, Access) {
 	if checker.Config.AccessCheckMode.IsReadableAccess(member.Access) ||
-		checker.containerTypes[member.ContainerType] {
+		// only allow references unrestricted access to members in their own container that are not entitled
+		// this prevents rights escalation attacks on entitlements
+		(member.Access.isPrimitiveAccess() && checker.containerTypes[member.ContainerType]) {
```

Comment:
This change is needed so that the following fails:

```
entitlement E
resource R {
	access(E) fun foo() {}
}
attachment Test for R {
	access(E) fun bar() {}
	fun foo(t: &Test) {
		t.bar()
	}
}
```

This test was added to the attachment test suite

Reply:
`t` has type `&Test`, which is unauthorized/unentitled. Why does the call `t.bar()` need to be checked specially?

Reply:
In the original logic, the `checker.containerTypes[member.ContainerType]` would be true, since the `t` has type `&Test`, and the container during the checking of that member access is `Test`. Without this change, we would thus permit the `t.bar()` call because the container is the defining attachment, even though `bar` requires `E`, which `t` doesn't have.